### PR TITLE
NCL-271 - add Json configs, modules, test

### PR DESCRIPTION
**Pull request by pslegr**
Matej, here is the #271 - please check... feel free to modify and let me know ;)

thanks!
Pavel

### CODE REVIEW

**Comment by slaskawi** [comment]
retest this please

File: common/pom.xml [hunk 1]
```diff
@@ -24,6 +24,18 @@
             <groupId>org.jboss</groupId>
             <artifactId>jboss-common-core</artifactId>
         </dependency>
+        
+        <dependency>
+          <groupId>com.fasterxml.jackson.core</groupId>
+          <artifactId>jackson-core</artifactId>
+          <version>2.4.4</version>
```
Comment (slaskawi) [inline common/pom.xml]: Versions should be in parent pom

File: common/src/main/java/org/jboss/pnc/common/Configuration.java [hunk 2]
```diff
@@ -24,12 +33,23 @@ public Configuration(final Properties properties) {
         this.properties = properties;
     }
 
-    //TODO return only part containing config for requested module
-    public Properties getModuleConfig(final String moduleTag) {
-        return properties;
+//    //TODO return only part containing config for requested module
```
Comment (slaskawi) [inline common/src/main/java/org/jboss/pnc/common/Configuration.java]: Let's delete that

File: common/src/main/java/org/jboss/pnc/common/json/ConfigurationJSONParser.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package org.jboss.pnc.common.json;
+
+import java.io.File;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+
+/**
```
Comment (slaskawi) [inline common/src/main/java/org/jboss/pnc/common/json/ConfigurationJSONParser.java]: Maybe we should remove it? Or fill in with some important data

**Comment by pslegr** [comment]
that's an old PR, lets close this one... as  another one has been merged already to master

File: common/src/main/java/org/jboss/pnc/common/json/ConfigurationJSONParser.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+public class ConfigurationJSONParser<T extends AbstractModuleConfig> {
+
+    public T parseJSONConfig(File configFile, Class<T> classType) throws ConfigurationParseException {
+        try {
+            // read from file, convert it to config module class
+            ObjectMapper mapper = new ObjectMapper();
+            ModuleConfigJson json_config = mapper.readValue(configFile, ModuleConfigJson.class);
+
+            for (AbstractModuleConfig config : json_config.getConfigs()) {
+              if(config.getClass().isAssignableFrom(classType)) {
```
Comment (slaskawi) [inline common/src/main/java/org/jboss/pnc/common/json/ConfigurationJSONParser.java]: We can also Cast without checking. If the type is incorrect we will end up in catch block and the resulting exception would be the same.

File: common/src/main/java/org/jboss/pnc/common/json/ConfigurationParseException.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package org.jboss.pnc.common.json;
+
+public class ConfigurationParseException extends Exception{
+    
+    /**
+     *
```
Comment (slaskawi) [inline common/src/main/java/org/jboss/pnc/common/json/ConfigurationParseException.java]: Empty Javadoc

File: common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            moduleConfigJson.addConfig(mavenRepoDriverModuleConfig);
+            
+            ObjectMapper mapper = new ObjectMapper();
+            mapper.writeValue(System.out, moduleConfigJson);
+            
+            //Assert.assertNotNull(jen_config);
+            //System.out.println(jen_config);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java]: Add Exception to throws clause and remove this `printStackTrace`.

File: common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                    new JenkinsBuildDriverModuleConfig(new URL("http://test/url"), "pavel", "test");
+            MavenRepoDriverModuleConfig mavenRepoDriverModuleConfig = 
+                    new MavenRepoDriverModuleConfig("http://something/base");
+            moduleConfigJson.addConfig(jenkinsBuildDriverModuleConfig);
+            moduleConfigJson.addConfig(mavenRepoDriverModuleConfig);
+            
+            ObjectMapper mapper = new ObjectMapper();
+            mapper.writeValue(System.out, moduleConfigJson);
+            
+            //Assert.assertNotNull(jen_config);
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java]: I think there should be some assertions here.

File: common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+import org.junit.Assert;
+import org.junit.Ignore;
+import org.junit.Test;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class ModuleConfigJsonTest {
+
+    @Test
+    public void serialize() {
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java]: We may change the name to something a little bit better:
* testSerializationWithSuccess
* shouldSucceedOnSerializingProperConfiguration etc

File: common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                    + ",\"baseUrl\":\"http://something/base\"}]}";    
+
+            ObjectMapper mapper = new ObjectMapper();
+            ModuleConfigJson config = mapper.readValue(json, ModuleConfigJson.class);
+            
+            Assert.assertNotNull(config);
+            System.out.println(">>> deserialize");
+            System.out.println(config);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/ModuleConfigJsonTest.java]: The same trick with catch here.

File: common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    public void test() {
+        try {
+            Configuration<JenkinsBuildDriverModuleConfig> configuration = 
+                    new Configuration<JenkinsBuildDriverModuleConfig>();
+            JenkinsBuildDriverModuleConfig jen_config = 
+                    configuration.getModuleConfig(JenkinsBuildDriverModuleConfig.class);
+            Assert.assertNotNull(jen_config);
+            System.out.println(jen_config);
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java]: And here ;)

File: common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Test
+    public void test() {
+        try {
+            Configuration<JenkinsBuildDriverModuleConfig> configuration = 
+                    new Configuration<JenkinsBuildDriverModuleConfig>();
+            JenkinsBuildDriverModuleConfig jen_config = 
+                    configuration.getModuleConfig(JenkinsBuildDriverModuleConfig.class);
+            Assert.assertNotNull(jen_config);
+            System.out.println(jen_config);
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java]: `System.out.println` might be removed.

File: common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+import org.jboss.pnc.common.Configuration;
+import org.jboss.pnc.common.json.ConfigurationParseException;
+import org.jboss.pnc.common.json.moduleconfig.JenkinsBuildDriverModuleConfig;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class JenkinsBuildDriverModuleConfigTest {
+
+    @Test
+    public void test() {
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/moduleconfig/JenkinsBuildDriverModuleConfigTest.java]: There must be a better name for this test

File: common/src/test/java/org/jboss/pnc/common/json/moduleconfig/MavenRepoDriverModuleConfigTest.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.io.IOException;
+
+import org.jboss.pnc.common.Configuration;
+import org.jboss.pnc.common.json.ConfigurationParseException;
+import org.jboss.pnc.common.json.moduleconfig.JenkinsBuildDriverModuleConfig;
+import org.jboss.pnc.common.json.moduleconfig.MavenRepoDriverModuleConfig;
+import org.junit.Assert;
+import org.junit.Test;
+
```
Comment (slaskawi) [inline common/src/test/java/org/jboss/pnc/common/json/moduleconfig/MavenRepoDriverModuleConfigTest.java]: The same applies to this class.

File: maven-repository-manager/src/main/java/org/jboss/pnc/mavenrepositorymanager/RepositoryManagerDriver.java [hunk 14]
```diff
@@ -64,16 +68,14 @@ public RepositoryConfiguration createRepository(ProjectBuildConfiguration projec
...
+        try {
+            MavenRepoDriverModuleConfig config = configuration.getModuleConfig(MavenRepoDriverModuleConfig.class);
+            String baseUrl = config.getBaseUrl();
             url = buildUrl(baseUrl, "api", "group", id);
-//        } catch (MalformedURLException e) {
-//            throw new RepositoryManagerException("Cannot format Maven repository URL. Base URL was: '%s'. Reason: %s", e,
-//                    baseUrl, e.getMessage());
-//        }
+        } catch (ConfigurationParseException e) {
+            e.printStackTrace();
```
Comment (slaskawi) [inline maven-repository-manager/src/main/java/org/jboss/pnc/mavenrepositorymanager/RepositoryManagerDriver.java]: Can we do something better with this exception? Maybe we could handle or rethrow/wrap it?

File: maven-repository-manager/src/test/java/org/jboss/pnc/mavenrepositorymanager/RepositoryManagerDriverTest.java [hunk 15]
```diff
@@ -1,28 +1,37 @@
...
     public void setup() {
-        Properties props = new Properties();
-        props.setProperty("base.url", "http://localhost/");
-        Configuration config = new Configuration(props);
+        Configuration<MavenRepoDriverModuleConfig> config = null;
+        try {
+            config = new Configuration<MavenRepoDriverModuleConfig>();
+        } catch (IOException e) {
+            // TODO Auto-generated catch block
+            e.printStackTrace();
```
Comment (slaskawi) [inline maven-repository-manager/src/test/java/org/jboss/pnc/mavenrepositorymanager/RepositoryManagerDriverTest.java]: `printStackTrace` to be removed.

**Pull request closed without merging**